doc/assist.md

```diff
@@ -0,0 +1,155 @@
+# ASSIST support
+
+ASSIST stands for **A**ssembler **S**ystem for **S**tudent **I**nstruction and **S**ystems **T**eaching.
+
+It is an IBM System/370-compatible assembler and interpreter developed in the early 1970s at Penn State 
+University by Graham Campbell and John Mashey.[^1]
+[^1]:See [ASSIST (computing) at Wikipedia](https://en.wikipedia.org/wiki/ASSIST_(computing)) for more details.
+
+The ASSIST support in z390 allows you to use the I/O and debugging instructions as described in the ASSIST
+User Manual [PART II. INPUT/OUTPUT AND DEBUGGING INSTRUCTIONS](https://web.archive.org/web/20150406052304/http://faculty.cs.niu.edu/~hutchins/csci640/asusergd.html#part2).
+
+## Using ASSIST instructions with z390
+
+### `ASSIST` option
+
+The ASSIST option must be provided to use the ASSIST instructions. See [z390 Options](Options/z390_options.md) for details.
+
+### `assist` command script
+
+z390 `assist` command script can be used to assemble, link, and execute ASSIST programs.
+
+``` dos
+assist {programname}
+```
+
+The `assist` command script enables ASSIST option and defines the following files for input and output:
+
+* XREAD={programname}.XRD
+* XPRNT={programname}.XPR
+* XPNCH={programname}.XPH
+* XGET={programname}.XGT
+* XPUT={programname}.XPT
+
+## File processing
+
+* Files are optional and will be opened on first access and closed at end of execution or end of file.
+* All files are ASCII with automatic conversion from or to EBCDIC. 
+* ASCII file records have trailing spaces removed and end with carriage return x'0d' and line feed x'0a'.
+* Input fields are padded with EBCDIC spaces x'40'.
+* Non ASCII output bytes are converted to periods.
+
+## Default ASSIST option behavior
+
+Using the ASSIST option enables z390 options NOLOADHIGH and NOINIT for compatibility with the original ASSIST assembler
+and simplifies relative address calculations.
+
+This has the following impacts:
+
+* Programs are loaded at X'10000' instead of high end of memory.
+* Registers are initialized to X'F4'.
+* Memory above the PSA is initialized to X'F5'.
+* Uninitialized areas of 390 load modules are initialized to X'F6'.
+
+You can use the options LOADHIGH and INIT to override this behavior. See [z390 Options](Options/z390_options.md) for more information.
+
+## Error handling
+
+A trace table is used to display the last 10 instructions prior to abnormal termination of a program when the TRACE option is off.
+
+The PSW displayed at abnormal termination includes ILC, CC, MASK, and AMODE.
+
+## ASSIST instructions
+
+### XDECI - Convert decimal to binary
+
+Format: `RX X'53rxbddd' r1.s2`
+
+* Start scan for next decimal number at s2 skipping leading blanks. 
+* Convert decimal number to binary in r1 until non decimal character found.
```

Comment:
I am assuming that is the behavior of z390 ASSIST as well.

Reply:
A quick test should do to remove any doubts as to actual behaviour of z390

Reply:
OK.. so I wrote a little program to check this... 

```
tstxdeci CSECT 
         BASR  15,0
         USING *,15
         SR    5,5   Clear R5
         XDECI 5,DECNUM
         BRC   b'1000',OK
         BRC   b'0001',ERR
         J     DUMP
OK       XPRNT =C'CC=0 OK',7
         J     DUMP
ERR      XPRNT =C'CC=3 ERR',8
DUMP     XDUMP 0
         BR    14
         DS    0F
DECNUM   DC   << value to check >>
         END
```

The original manual provides the following:

> XDECI is generally used to scan a data card read by XREAD. The sequence of actions performed by XDECI is as follows:
> 
> 1. Beginning at the location given by ADDRESS, memory is scanned for the first character which is not a blank.
> 2. If the first character found is anything but a decimal digit or plus or minus sign, register 1 is set to the address of that character, and the condition code is set to 3 (overflow) to show that no decimal number could be converted. The contents of REG are not changed, and nothing more is done.
> 3. From one to nine decimal digits are scanned, and the number converted to binary and placed in REG, with the appropriate sign. The condition code is set to 0 (0), 1 (-), or 2 (+), depending on the value just placed in REG.
> 4. Register 1 is set to the address of the first non-digit after the string of decimal digits. Thus REG should not usually be 1. This permits the user to scan across a card image for any number of decimal values. The values should be separated by blanks, since otherwise the scanner could hang up on a string like -123*, unless the user checks for this himself. I.e. XDECI will skip leading blanks but will not itself skip over any other characters.
> 5. During step 3, if ten or more decimal digits are found, register 1 is set to the address of the first character found which is not a decimal digit, the condition code is set to 3, and REG is left unchanged. A plus or minus sign alone causes a similar action, with R1 set to the address of the character following the sign character.
> 

**Value to check = P'1234'**
Sets CC=3. R5=0

**Value to check = Z'1234'**
Sets CC=3. R5=7B/123 
Does not accept signed zoned decimal..

**Value to check = X'F1F2F3F4'**
Sets CC=0, R5=4D2/1234

**Value to check = Z'1234567890'**
Sets CC=0. R5=075BCD15/123456789

So my findings are:
* Does not deal with packed decimal at all (I think this is intentional)
* Does not deal with signed zoned decimal (ASSIST doc says XDECI can be used to scan input cards for signed or unsigned decimal numbers)
* CC=3 is being used to suggest invalid input (no valid chars)
* CC=3 is not being used when the input exceeds 9 digits. Translates up to 9, still returns CC=0.
* Also did not see it setting the CC based on the value (positive, neg, zero)

I would suggest that our XDECI implementation is not the same as the original ASSIST..
This could be a bug, a partial implementation or by design.

I am just writing the doco so I need some more details to finalise..

@dsh33782 @abekornelis @jyganci Your feedback and next steps on the above would be appreciated.

Reply:
Anthony, good find. Would you please create an issue to cover your findings? I'd expect we can fix it relatively quickly. The program you wrote could be used for RT purposes. If you add it to the issue, we don't need to rewrite it ;-)

Reply:
I have raised issue #291 to cover the deviation in behaviour between z390 and original. I'm going to close this off at this point and any doc changes can be done later.